Display SVG files as images instead of text

jtran opened the pull request:
This is an attempt to close #1095.  It makes the following changes:

- Detects SVG image files and serves them with `Content-Type: image/svg+xml` instead of `text/plain`, `text/html`, or `text/xml`.  This allows users to view an SVG as an image when browsing a repo's tree and when it's embedded in a markdown file.
- Serves extra security headers for SVG files in raw and media routes: `Content-Security-Policy: default-src 'none'; style-src 'unsafe-inline'; sandbox` and `X-Content-Type-Options: nosniff`.  This prevents the execution of scripts in untrusted SVG files.
- Adds regression tests for the above
- UI toggle on the view file page to switch between the text source and the rendered image for SVG files
- Adds a config setting to disable the feature
- ~Hides the "raw" button when viewing SVG files since unsafe styles aren't displayed, and viewing the image without styles may be confusing.  Users can always right-click on the image to view or save it.~

I tested this with https://github.com/edgar-bonet/test-svg-mime in Firefox and Chrome.

techknowlogick commented on integrations/view_test.go:
```diff
@@ -0,0 +1,26 @@
+// Copyright 2017 The Gitea Authors. All rights reserved.
```
Could you update the year to 2020?

techknowlogick commented on custom/conf/app.example.ini:
```diff
@@ -258,6 +258,10 @@ CUSTOM_URL_SCHEMES =
 ; Separate the extensions with a comma. To render files without any extension as markdown, just put a comma
 FILE_EXTENSIONS = .md,.markdown,.mdown,.mkd
 
+[svg_render]
```
Could this be grouped with other ui settings (eg. `[ui.svg]`)?

techknowlogick replied:
Another note, could the setting itself be `ENABLE_RENDER`?

techknowlogick replied:
Once the above have been modified, could you also add the additional settings to the "cheat sheet" found in docs/ directory.

techknowlogick reviewed (commented):
Thank you for this PR, and especially thank you for including tests & documentation :)

This is not an exhaustive review, just a few preliminary comments.

silverwind commented:
> Hides the "raw" button when viewing SVG files since unsafe styles aren't displayed, and viewing the image without styles may be confusing. Users can always right-click on the image to view or save it.

I have to say I like the "Raw" button on GitHub for SVG. It just opens the SVG with the proper mime type in a new tab. Which styles are you refering to? SVGs are self-contained.

jtran commented:
@silverwind,

Without adding `style-src 'unsafe-inline';` to the content-security-policy, I haven't been able to view styles that are contained within SVG images.  When you view an image directly by navigating to the raw URL in your browser, `<img>` tag security measures do not apply.  So in this case, the content-security-policy is the only thing preventing malicious scripts inside SVGs from executing.  Apparently there are also potentially harmful styles too?  But I'm not an expert on this.

If someone understands this issue better and can explain why adding `style-src 'unsafe-inline';` would be safe without additional security measures (like a separate domain), I would happily add that.  But I'm not sure about the implications.

silverwind commented:
Good read on SVG security: https://www.w3.org/wiki/SVG_Security

You're right, `<img>` seems to be save to do while serving raw with `image/svg+xml` will execute scripts, so I guess we'd need to sanitize the SVG first for this case (stripping script tags and `url` references).

jtran commented:
> serving raw with image/svg+xml will execute scripts, so I guess we'd need to sanitize the SVG first for this case

That's why we use `Content-Security-Policy: default-src 'none'; sandbox`, which disables scripts.  Sanitizing isn't needed to disable scripts.  Please read up on #1095.  Although they aren't mutually exclusive, my understanding from that issue is that people favor sandboxing over filtering/sanitizing.  A sanitizing pull request #8024 was abandoned.

silverwind commented:
Few minor UI things:

1. Display lines of code for SVG, e.g. like GitHub:

<img width="277" alt="image" src="https://user-images.githubusercontent.com/115237/102935732-05a72780-44a7-11eb-821d-8c6432fd80cd.png">

2. Editing should be allowed but UI currently shows "Cannot edit binary files"
3. Also use the rendered view in diffs, currently it's showing source.

I think ultimately we'll need a switchable view with source and rendered but I guess rendered only will work as a start.

silverwind commented:
Regarding raw: How about just showing the source in `text/plain`? I think it makes more sense, I'm usually after the source and currently there's no direct way to show it.

I also noticed SMIL animations ([example](https://raw.githubusercontent.com/StylishThemes/GitHub-Dark/e50dcf2b334d5fc7e6f7654e73ad4ad05d57d0fa/images/octocat-spinner-smil.svg)) do not play when right clicking and `view image`. Is that caused by the sandboxing?

silverwind commented on routers/repo/download.go:
```diff
@@ -46,6 +46,11 @@ func ServeData(ctx *context.Context, name string, reader io.Reader) error {
 		ctx.Resp.Header().Set("Content-Disposition", fmt.Sprintf(`inline; filename="%s"`, name))
 		ctx.Resp.Header().Set("Access-Control-Expose-Headers", "Content-Disposition")
+		if base.IsSVGImageFile(buf) {
+			ctx.Resp.Header().Set("Content-Security-Policy", "default-src 'none'; sandbox")
```
```suggestion
			ctx.Resp.Header().Set("Content-Security-Policy", "default-src 'none'; style-src 'unsafe-inline'; sandbox")
```

This fixes the SMIL animation in Firefox (which counts as inline style according to [this](https://bugzilla.mozilla.org/show_bug.cgi?id=763879)) and matches the CSP GitHub uses.

silverwind commented on custom/conf/app.example.ini:
```diff
@@ -242,6 +242,10 @@ TIMEOUT_STEP = 10s
 EVENT_SOURCE_UPDATE_TIME = 10s
 
+[ui.svg]
+; Enable rendering of SVG files as images instead of as text.
```
```suggestion
; Whether to render SVG files as images.  If SVG rendering is disabled, SVG files are displayed as text and cannot be embedded in markdown files as images.
```

We usually try to keep these in sync.

kdumontnu commented:
> Regarding raw: How about just showing the source in `text/plain`? I think it makes more sense, I'm usually after the source and currently there's no direct way to show it.

Agree with `text/plain` view for raw. I've started work on a toggle button for regular file view. I'll continue that once this is merged

jtran commented:
> Regarding raw: How about just showing the source in text/plain? I think it makes more sense, I'm usually after the source and currently there's no direct way to show it.

If we did that, then embedding SVG images in markdown wouldn't work.  Assuming the embedding uses the same raw URL, the server cannot distinguish between the two cases.

> I also noticed SMIL animations ([example](https://raw.githubusercontent.com/StylishThemes/GitHub-Dark/e50dcf2b334d5fc7e6f7654e73ad4ad05d57d0fa/images/octocat-spinner-smil.svg)) do not play in Firefox (they do in Chrome) when right clicking and view image. Is that caused by the sandboxing?

The animation plays for me in Firefox v84.0.1 when I follow that link.

silverwind commented:
> The animation plays for me in Firefox v84.0.1 when I follow that link.

Yes they play in GitHub but not in Gitea because of the missing `style-src 'unsafe-inline'` directive that GitHub serves.

> If we did that, then embedding SVG images in markdown wouldn't work. Assuming the embedding uses the same raw URL, the server cannot distinguish between the two cases.

I think it's unavoidable we add a JS-switchable view for rendered/source mode. Do you think you can implement something like that here?

jtran commented:
Yes, I can add a way in the UI to toggle between text source and image.

Should I add `style-src 'unsafe-inline'` also?  I just didn't want to add it if there was a security risk.

jtran commented:
To clarify, it's not clear to me that we can just do what GitHub does because GitHub has additional security measures like serving content from a different domain.  Presumably, serving from another domain is a much larger change that is out of scope for this pull request.

silverwind commented:
> Should I add style-src 'unsafe-inline' also? I just didn't want to add it if there was a security risk.

It's the only way to get SMIL to work in Firefox and I think it's an essential part of SVG. I don't see anything wrong allowing inline styles. I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1683972 because I believe classifying SMIL as inline style is wrong in Firefox.

jtran commented:
I think I addressed all the comments.  This PR doesn't affect the pull request or commit diff view.  Changing the diff display and adding JS toggling to it seems a bit more involved than I was hoping.  Can this be merged without that?

Thank you for the timely review.  This is my first contribution to the project.

silverwind commented:
I agree, JS toggle can come later.

kdumontnu commented:
@silverwind @techknowlogick @zeripath If you have a chance can you take a look through this PR? Should be all set & a big improvement to how gitea handles svgs. Most of the files in the diff are for integration test.

lunny commented on routers/repo/view.go:
```diff
@@ -396,6 +396,20 @@ func renderFile(ctx *context.Context, entry *git.TreeEntry, treeLink, rawLink st
+	ctx.Data["IsDisplayingRendered"] = isDisplayingRendered
+	isRepresentableAsText := base.IsRepresentableAsText(buf)
+	ctx.Data["IsRepresentableAsText"] = isRepresentableAsText
+	if !isRepresentableAsText {
```
So we should also change `ctx.Data["IsDisplayingSource"]`?

jtran replied:
Fixed.

lunny commented:
Unfortunately, it seems it's not work. I migrated from  https://github.com/edgar-bonet/test-svg-mime to local Gitea instance.

lunny commented on routers/repo/download.go:
```diff
@@ -46,6 +46,11 @@ func ServeData(ctx *context.Context, name string, reader io.Reader) error {
 		ctx.Resp.Header().Set("Content-Disposition", fmt.Sprintf(`inline; filename="%s"`, name))
 		ctx.Resp.Header().Set("Access-Control-Expose-Headers", "Content-Disposition")
+		if base.IsSVGImageFile(buf) {
+			ctx.Resp.Header().Set("Content-Security-Policy", "default-src 'none'; style-src 'unsafe-inline'; sandbox")
```
Is `unsafe-inline` necessary here? ref. https://content-security-policy.com/unsafe-inline/

jtran replied:
Without it, many SVG images look nothing like they're intended to look.  Many SVG attributes have a style alternative.

That said, I originally omitted it, but I was asked to add it in another comment in this PR.  As long as it's safe, I think that including it is a good thing.

lunny replied:
Please see my ref article, It said
>
When you want to allow inline scripts or styles on a page that uses CSP, there two much better options: nonce or hash.

jtran replied:
I read the article.  I'm having trouble understanding how nonce or hash is actually better.  If I understand it correctly, a hash prevents modification between the server and client, which is not more than what HTTPS would do.  So this is to make the HTTP case safer.

If you think that's what we should do, it's not clear to me how to actually implement it.  It seems like we would need to parse the SVG file enough to either insert nonces at the appropriate places in the SVG or read all the inline styles so that we can compute their hashes.  Is this what you're saying?  I'm guessing this must include all the `style="..."` attributes.  There could potentially be many of them.  I'll think about it some more.

CirnoT replied:
@lunny The suggestion from the ref article is not applicable in this case, as the CSP is applied when rendering user-generated content (the SVG file). The general purpose of using nonce is so that website can include inline script in non-user controlled HTML content while at the same time benefiting from CSP blocking possible XSS attacks.

lunny replied:
nonce of CSP requires a change to the svg content dynamiclly which will make the thing complicated.

kdumontnu commented:
Is the "View Raw" case really commonly used? In my opinion we should take the most conservative/secure option for view raw or disable it for SVG and keep the size of this PR small. 
This PR already fixes embedding svgs in markdown, renders in view-file, and toggles between render/code, which is a huge improvement. If the view raw case isn't perfect, it can be discussed and improved in a subsequent PR.

zeripath commented on modules/base/tool.go:
```diff
@@ -28,6 +29,15 @@ import (
 )
 
+// Use at most this many bytes to determine Content Type.
+const sniffLen = 512
```
mostly we use 1024 bytes - I think that's what works best for most external libraries too.

zeripath replied:
would it be possible to add a test to ensure that if a file contains SVG beyond the sniff length that the SVG is still stripped? I could imagine it being something that unfamiliar developers might want to remove and we must not allow it to sneak back in unsafely.

jtran replied:
I copied this from the [`go http module`](https://github.com/golang/go/blob/dev.boringcrypto.go1.15/src/net/http/sniff.go#L13) that was being used before.  I can add a test, but I'm pretty sure 512 is what was being used before.

zeripath replied:
I stand corrected! I'm misremembering the bytes/runes thing and lfs etc.

Would you be able to add a test that ensures a file with the svg tag after 512 chars still gets sanitized?

zeripath replied:
(I know your code doesn't change this - it's protecting against someone else removing the sanitization later by mistake

jtran replied:
I'd be happy to add a test, but to be clear, there is no sanitizing or stripping in this PR.  The security is by the nature of how `<img>` tags and content-security-policies work.  So I'm not sure what the test would actually assert.  If the svg tag weren't in the first 512 bytes, the file would get detected as something else, such as text/plain, text/html, or text/xml and served as that.  When viewing the file, gitea would display it as a text file.  When viewing the raw file, the browser will render it as text.  When embedded in an `<img>` tag from a markdown file, the image would fail to display.

jtran replied:
BTW, as it's currently implemented, we only attempt to look for an SVG content type if it's already been detected as one of: text/plain, text/html, text/xml.  So it's not possible that having the svg tag later in the file than 512 bytes would cause a file to be detected as something other than those three content types.

zeripath replied:
That's exactly my point - we want the HTML file with embedded SVG after 512 to have the SVG sanitised away and displayed without it.

The test would simply assert that there were no SVG els.

The point is to prevent some well meaning change to the sanitizer leading to unsanitised SVG leaking.

zeripath replied:
If you really don't want to do it - it can be added later but it'd be good to get the test in as an example of something that is not going to be supported and also something we don't want to happen.

jtran replied:
I think there's a misunderstanding.  Maybe I'm not being clear enough.  If there's a file that's detected as text/html with a root `<svg>` tag after 512 bytes, and we serve it with `Content-Type: text/html`, the browser will never up-convert to `image/svg+xml` because we're also  serving the header [`X-Content-Type-Options: nosniff`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options), which tells the browser, "take our word for it; don't try to do your own sniffing".

So I don't see a need to assert that there are no `<svg>` tags.  Am I missing something?

jtran replied:
I'm working on a test that asserts that a file with a root `<svg>` tag after 512 bytes is detected as `text/html`.  Hopefully, that's what you mean.

vmario89 commented:
hi. saw that this was merged but it was not yet released right? Tried to configure Gitea ui.svg ENABLE_RENDER = true in app.ini but i guess thats another setting, not related to this? Looking forward to directly display SVG files in preview instead of plaintext.


    https://github.com/go-gitea/gitea/issues/1095
    https://github.com/go-gitea/gitea/pull/14101
    https://docs.gitea.io/en-us/config-cheat-sheet/#ui---svg-images-uisvg

kdumontnu commented:
> hi. saw that this was merged but it was not yet released right? Tried to configure Gitea ui.svg ENABLE_RENDER = true in app.ini but i guess thats another setting, not related to this? Looking forward to directly display SVG files in preview instead of plaintext.

I believe it is queued for release in version [`1.14`](https://github.com/go-gitea/gitea/milestone/76). You can test it in the current master branch.
The setting is enabled by default, so no need to change `app.ini`
